Code/Editor/Lib/Tests/test_ModularViewportCameraController.cpp

```diff
@@ -0,0 +1,165 @@
+        QObject::connect(
+            m_inputChannelMapper.get(), &AzToolsFramework::QtEventToAzInputMapper::InputChannelUpdated, m_rootWidget.get(),
+            [this, nativeWindowHandle](const AzFramework::InputChannel* inputChannel, [[maybe_unused]] QEvent* event)
+            {
+                m_controllerList->HandleInputChannelEvent(
+                    AzFramework::ViewportControllerInputEvent{ TestViewportId, nativeWindowHandle, *inputChannel });
+            });
+
+        using ::testing::NiceMock;
+        using ::testing::Return;
+
+        NiceMock<MockWindowRequests> mockWindowRequests;
+        mockWindowRequests.Connect(nativeWindowHandle);
+
+        // note: WindowRequests is used internally by ModularViewportCameraController, this ensures it returns the viewport size we want
+        ON_CALL(mockWindowRequests, GetClientAreaSize())
+            .WillByDefault(Return(AzFramework::WindowSize(WidgetSize.width(), WidgetSize.height())));
+
+        // create editor modular camera
+        auto controller = CreateModularViewportCameraController(TestViewportId);
+
+        // set some overrides for the test
+        AtomToolsFramework::ModularCameraViewportContext* cameraViewportContextView = nullptr;
+        controller->SetCameraViewportContextBuilderCallback(
+            [&cameraViewportContextView](AZStd::unique_ptr<AtomToolsFramework::ModularCameraViewportContext>& cameraViewportContext)
+            {
+                cameraViewportContext = AZStd::make_unique<TestModularCameraViewportContextImpl>();
+                cameraViewportContextView = cameraViewportContext.get();
+            });
+
+        controller->SetCameraPropsBuilderCallback(
+            [](AzFramework::CameraProps& cameraProps)
+            {
+                cameraProps.m_rotateSmoothingEnabledFn = []
+                {
+                    return false;
+                };
+
+                cameraProps.m_translateSmoothingEnabledFn = []
+                {
+                    return false;
+                };
+            });
+
+        m_controllerList->Add(controller);
+
+        // move to the center of the screen
+        auto start = QPoint(WidgetSize.width() / 2, WidgetSize.height() / 2);
+        MouseMove(m_rootWidget.get(), start, QPoint(0, 0));
+        m_controllerList->UpdateViewport({ TestViewportId, AzFramework::FloatSeconds(0.01666f), AZ::ScriptTimePoint() });
+
+        // When
+        // move mouse diagonally to top right, then to bottom left and back repeatedly
+        auto current = start;
+        auto halfDelta = QPoint(200, -200);
+        for (int diagonals = 0; diagonals < 80; ++diagonals)
+        {
+            for (int i = 0; i < 50; ++i)
+            {
+                MousePressAndMove(m_rootWidget.get(), current, halfDelta / 50, Qt::MouseButton::RightButton);
+                m_controllerList->UpdateViewport({ TestViewportId, AzFramework::FloatSeconds(0.01666f), AZ::ScriptTimePoint() });
```

Comment:
Would an additional test that uses differing framerates be useful?  I don't mean random frametimes, but some preset array of frametimes that are not exactly 60fps that is looped through on each mouse move

Reply:
That's an interesting idea, it shouldn't be too hard to make this a parameterized test and have it vary based on `deltaTime` - I think I'd rather do that than just add a separate test but it might be better to do that as a separate PR. I'll do a fast follow-up with that transformation so we can see the changes more clearly in the diff (as right now this file is an add).